### PR TITLE
[WIP] Program indicators cohort support

### PR DESCRIPTION
**Pull request by Bekkalizer**

### CODE REVIEW

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/program/AnalyticsPeriodBoundary.java [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+    public static final String DB_INCIDENT_DATE = "incidentdate";
+    
+    
+    private String boundaryTarget;
+    
+    private AnalyticsPeriodBoundaryType analyticsPeriodBoundaryType;
+    
+    private PeriodType offsetPeriodType;
+    
+    private Integer offsetNumberOfPeriods;
```
Comment (larshelge) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/program/AnalyticsPeriodBoundary.java]: Rename to "offsetPeriods", number of is implicit as Integer

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/DefaultEventQueryPlanner.java [hunk 2]
```diff
@@ -242,7 +244,10 @@ else if ( params.isCollapseDataDimensions() && !params.getItems().isEmpty() )
     {
         List<EventQueryParams> queries = new ArrayList<>();
         
-        if ( params.getAggregationTypeFallback().isLastPeriodAggregationType() && !params.getPeriods().isEmpty() )
+        
+        if ( ( params.getAggregationType() != null && params.getAggregationType().isLastPeriodAggregationType() ) ||
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/DefaultEventQueryPlanner.java]: Add utility methods on DataQueryParams to remove 2 of these conditions, makes it more readable.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/program/ProgramIndicator.java [hunk 3]
```diff
@@ -264,6 +274,43 @@ public void updateIndicatorGroups( Set<ProgramIndicatorGroup> updates )
...
+        return this.analyticsType == AnalyticsType.EVENT && !this.analyticsPeriodBoundaries.equals( defaultEventTypeBoundaries ) ||
+            this.analyticsType == AnalyticsType.ENROLLMENT && !this.analyticsPeriodBoundaries.equals( defaultErollmentTypeBoundaries );
+    }
+    
+    public Boolean hasEventBoundary()
+    {
+        return getEndEventBoundary() != null || getStartEventBoundary() != null;
+    }
+    
+    public AnalyticsPeriodBoundary getEndEventBoundary()
```
Comment (larshelge) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/program/ProgramIndicator.java]: Javadocs

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/program/AnalyticsPeriodBoundary.java [hunk 4]
```diff
@@ -0,0 +1,251 @@
+package org.hisp.dhis.program;
+
+import java.text.SimpleDateFormat;
```
Comment (larshelge) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/program/AnalyticsPeriodBoundary.java]: Imports after license

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/JdbcEnrollmentAnalyticsManager.java [hunk 5]
```diff
@@ -268,7 +297,15 @@ else if ( params.hasEnrollmentProgramIndicatorDimension() )
         
         for ( DimensionalObject dimension : params.getDimensions() )
         {
-            columns.add( statementBuilder.columnQuote( dimension.getDimensionName() ) );
+            //When creating SQL for a program indicator with non-default boundaries, only one period is queried at a time
+            //this period needs to be hard-coded in the SQL return statement, as non-default boundaries is relative to the 
+            //original period.
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/JdbcEnrollmentAnalyticsManager.java]: Remove long inline comments, move to proper Javadoc on method.

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/JdbcEnrollmentAnalyticsManager.java [hunk 6]
```diff
@@ -304,21 +341,55 @@ else if ( ValueType.COORDINATE == queryItem.getValueType() )
...
-        
-        if ( params.hasStartEndDate() )
-        {        
-            sql += "where enrollmentdate >= '" + getMediumDateString( params.getStartDate() ) + "' ";
-            sql += "and enrollmentdate <= '" + getMediumDateString( params.getEndDate() ) + "' ";
+        if ( params.hasProgramIndicatorDimension() && params.getProgramIndicator().hasNonDefaultBoundaries() )
+        {
+            //The program indicator has non-default boundaries, and defines its own relationship with the 
+            //reporting period. We need to make custom where-clauses instead of using the preaggregated period columns.
+            //We know that the query planner has split the query into individual periods, as this is always done for
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/JdbcEnrollmentAnalyticsManager.java]: Remove long inline comments, move to proper Javadoc on method.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/DefaultProgramIndicatorService.java [hunk 7]
```diff
@@ -72,34 +72,30 @@
...
+        .put( ConditionalSqlFunction.KEY, new ConditionalSqlFunction() )
+        .put( HasValueSqlFunction.KEY, new HasValueSqlFunction() ).build();
+
+    private static final Map<String, String> VARIABLE_SAMPLE_VALUE_MAP = ImmutableMap.<String, String> builder()
+        .put( ProgramIndicator.VAR_COMPLETED_DATE, "'2017-07-08'" )
+        .put( ProgramIndicator.VAR_CURRENT_DATE, "'2017-07-08'" ).put( ProgramIndicator.VAR_DUE_DATE, "'2017-07-08'" )
+        .put( ProgramIndicator.VAR_ENROLLMENT_COUNT, "1" ).put( ProgramIndicator.VAR_ENROLLMENT_DATE, "'2017-07-08'" )
+        .put( ProgramIndicator.VAR_ENROLLMENT_STATUS, "'COMPLETED'" ).put( ProgramIndicator.VAR_EVENT_COUNT, "1" )
+        .put( ProgramIndicator.VAR_EVENT_DATE, "'2017-07-08'" )
+        .put( ProgramIndicator.VAR_EXECUTION_DATE, "'2017-07-08'" )
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/DefaultProgramIndicatorService.java]: Is this production code or test code?

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/DefaultProgramIndicatorService.java [hunk 8]
```diff
@@ -436,7 +431,36 @@ else if ( ProgramIndicator.KEY_CONSTANT.equals( key ) )
...
+    private String getDataValueEnrollmentSql( String dataElementUid, String programStageUid,
+        ProgramIndicator programIndicator, Date reportingStartDate, Date reportingEndDate )
+    {
+        if ( programIndicator.hasNonDefaultBoundaries() && programIndicator.hasEventBoundary() )
+        {
+            String eventTableName = "analytics_event_" + programIndicator.getProgram().getUid();
+            String columnName = "\"" + dataElementUid + "\"";
+            return "(select "
+                + columnName
+                + " from "
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/DefaultProgramIndicatorService.java]: Flatten out lines a bit

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/metadata/objectbundle/hooks/EmbeddedObjectObjectBundleHook.java [hunk 9]
```diff
@@ -96,25 +97,40 @@
...
+    }
+    
+    private void handleProperty( Object o, ObjectBundle bundle, Property property ) 
+    {
+        if ( property.isIdentifiableObject() )
+        {
+            ((BaseIdentifiableObject) o).setAutoFields();
+        }
+        
+        Schema embeddedSchema = schemaService.getDynamicSchema( o.getClass() );
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/metadata/objectbundle/hooks/EmbeddedObjectObjectBundleHook.java]: Perhaps ask @mortenoh for a review of this class.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/DefaultProgramIndicatorService.java (hunk 7 above)
Comment (Bekkalizer) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/DefaultProgramIndicatorService.java]: This is sample values that is used for validating expression format.

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/JdbcEnrollmentAnalyticsManager.java [hunk 10]
```diff
@@ -28,297 +28,103 @@
...
-            
-            if ( params.isIncludeNumDen() )
-            {
-                grid.addNullValues( 3 );
+                sql += sqlHelper.whereAnd() + "( select count * from analytics_event_" + params.getProgramIndicator().getProgram().getUid() + 
+                    " where pi = enrollmenttable.pi " + 
+                    (params.getProgramIndicator().getEndEventBoundary() != null ? 
+                    ( sqlHelper.whereAnd() + " " + 
+                    params.getProgramIndicator().getEndEventBoundary().getSqlCondition( params.getEarliestStartDate(), params.getLatestEndDate() ) + " ") 
+                    : "") +
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/JdbcEnrollmentAnalyticsManager.java]: This part could use some better formatting

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/JdbcEnrollmentAnalyticsManager.java [hunk 11]
```diff
@@ -28,297 +28,103 @@
...
-        SqlRowSet rowSet = jdbcTemplate.queryForRowSet( sql );
-
-        while ( rowSet.next() )
-        {            
-            grid.addRow();
-
-            if ( params.isAggregateData() )
+            //The program indicator has non-default boundaries, and defines its own relationship with the 
+            //reporting period. We need to make custom where-clauses instead of using the preaggregated period columns.
+            //We know that the query planner has split the query into individual periods, as this is always done for
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/JdbcEnrollmentAnalyticsManager.java]: Move long comment out of code into Javadoc on method

File: dhis-2/dhis-services/dhis-service-core/src/main/resources/org/hisp/dhis/program/hibernate/AnalyticsPeriodBoundary.hbm.xml [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+  <class name="org.hisp.dhis.program.AnalyticsPeriodBoundary" table="analyticsperiodboundary">
+
+    <cache usage="read-write" />
+
+    <id name="id" column="analyticsperiodboundaryid">
+      <generator class="native" />
+    </id>
+    &identifiableProperties;
+
+	<property name="boundaryTarget" length="50" />
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-core/src/main/resources/org/hisp/dhis/program/hibernate/AnalyticsPeriodBoundary.hbm.xml]: Codestyle, indent

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/AbstractJdbcEventAnalyticsManager.java [hunk 13]
```diff
@@ -0,0 +1,403 @@
...
+    {
+        List<String> columns = Lists.newArrayList();
+        
+        for ( DimensionalObject dimension : params.getDimensions() )
+        {
+            if ( dimension.getDimensionType() != DimensionType.PERIOD || !params.hasNonDefaultBoundaries() )
+            {
+                columns.add( statementBuilder.columnQuote( dimension.getDimensionName() ) );
+            }
+            else if ( params.getPeriods().size() == 1 )
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/AbstractJdbcEventAnalyticsManager.java]: Use explicit convenience methods, params.hasSinglePeriod, makes code more expressive

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/AbstractJdbcEventAnalyticsManager.java [hunk 14]
```diff
@@ -0,0 +1,403 @@
...
+            {
+                Period period = (Period) params.getPeriods().get( 0 );
+                columns.add( ( columnNamesOnly ? "" : ( period.getIsoDate() + " as " ) ) +
+                    period.getPeriodType().getName() );
+            }
+            else if ( params.getPeriods().size() == 0 && params.getFilterPeriods().size() > 0 )
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/AbstractJdbcEventAnalyticsManager.java]: Use !params.hasPeriods(), params.hasFilterPeriods() instead

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/AbstractJdbcEventAnalyticsManager.java [hunk 15]
```diff
@@ -0,0 +1,403 @@
...
+        sql += getFromClause( params );
+        
+        sql += getWhereClause( params );
+
+        // ---------------------------------------------------------------------
+        // Group by
+        // ---------------------------------------------------------------------
+        
+        List<String> selectColumnNames = getSelectColumnNames( params );
+        if ( selectColumnNames.size() > 0 )
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/AbstractJdbcEventAnalyticsManager.java]: Empty line between

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/AbstractJdbcEventAnalyticsManager.java [hunk 16]
```diff
@@ -0,0 +1,403 @@
...
+        }
+        else
+        {
+            itemValue = StringUtils.defaultString( itemValue, NA );
+            
+            return value + itemValue;
+        }
+    }
+    
+    protected abstract String getFromClause( EventQueryParams params );
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/AbstractJdbcEventAnalyticsManager.java]: Add javadoc to abstract methods.

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/JdbcEventAnalyticsManager.java [hunk 17]
```diff
@@ -544,22 +301,36 @@ private String getFromClause( EventQueryParams params )
...
+        SqlHelper sqlHelper = new SqlHelper();
         
         // ---------------------------------------------------------------------
         // Periods
         // ---------------------------------------------------------------------
-
-        if ( params.hasStartEndDate() )
+        if ( params.hasNonDefaultBoundaries() )
+        {
+            //The program indicator has non-default boundaries, and defines its own relationship with the
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/JdbcEventAnalyticsManager.java]: Move long comments out of code and into Javadoc on method

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/DefaultProgramIndicatorService.java [hunk 18]
```diff
@@ -436,7 +431,28 @@ else if ( ProgramIndicator.KEY_CONSTANT.equals( key ) )
...
 
-    @Override
+    private String getDataValueEnrollmentSql( String dataElementUid, String programStageUid,
+        ProgramIndicator programIndicator, Date reportingStartDate, Date reportingEndDate )
+    {
+        if ( programIndicator.hasNonDefaultBoundaries() && programIndicator.hasEventBoundary() )
+        {
+            String eventTableName = "analytics_event_" + programIndicator.getProgram().getUid();
+            String columnName = "\"" + dataElementUid + "\"";
+            return "(select " + columnName + " from " + eventTableName +
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/DefaultProgramIndicatorService.java]: Could have better formatting

**Pull request merged**